src/pudl/constants.py

```diff
@@ -312,8 +312,8 @@ class Partition(TypedDict, total=False):
             categories=EPACEMS_MEASUREMENT_CODES
         ),
         'heat_content_mmbtu': "float32",
-        'facility_id': pd.Int32Dtype(),  # Nullable Integer
-        'unit_id_epa': pd.Int32Dtype(),  # Nullable Integer
+        'facility_id': pd.Int32Dtype(),
+        'unit_id_epa': pd.StringDtype(),
```

Comment:
IIRC the only reason we haven't transitioned yet is because there's some metadata described for the EIA 861 that doesn't exist anywhere else. There are hundreds of columns that need to have metadata compiled for them, and probably some column renaming / simplification to do, and it's going to be a bit of a chore. And then we'll have to go through and replace all of the dataframe typing that happens everywhere, and make it use the new system.

I've been thinking about making a PUDL wrapper for `pd.read_sql()` that uses the table name to lookup the column data types and applies them automatically before returning the dataframe. Is there a more sensible way to do this? Applying the types by hand all the time seems kind of dumb / error prone.

Reply:
> I've been thinking about making a PUDL wrapper for `pd.read_sql()` that uses the table name to lookup the column data types and applies them automatically before returning the dataframe. Is there a more sensible way to do this? Applying the types by hand all the time seems kind of dumb / error prone.

This sounds like a good candidate for a `Resource` method, e.g. `Resource.read_df_from_sql()`, although it would mostly be `df = pd.read_sql(...)` followed by the existing `Resource.format_df(df)`. Since the primary key field handling in the latter may not apply in all situations, it may eventually need some switch arguments or to be split into two separate methods.

Reply:
I'm going to move on to compiling the EIA-861 metadata and hopefully make this transition. See also Issue #818.